### ISSUE TITLE
patchGlobal recursion protection can cause premature exit

### ISSUE THREAD

**Issue by pkaminski** [getsentry/raven-node#127]
I've run out of leads on this issue, so I'm hoping someone can give me a tip:

I have raven set up to capture global exceptions and terminate the process once the exception has been logged.  This works fine in development, and worked fine on Heroku, but doesn't capture the exception now that I'm running on Google Compute Engine (strictly speaking Managed VMs, a part of App Engine).  Exceptions captured explicitly with `captureError` work in all environments.  Here's how I set up the logging:

``` javascript
global.PROD = process.env.NODE_ENV === 'production';

// Normally only capture in production, but I removed the condition to test in dev.
global.raven = new (require('raven').Client)(PROD && ('https://<big_secret>@app.getsentry.com/<project>'));

process.on('uncaughtException', function(error) {
  console.log('Uncaught top-level exception!');
  console.log(error.stack);
  if (!PROD) process.exit(1);
});

raven.patchGlobal(function() {process.exit(1);});

// Intercept patches the function to invoke the callback before calling the original.
intercept(raven, ['captureMessage'], function(message) {
  console.log(message);
});
```

On GCE, I see the error being logged twice to the console and the VM restarting itself, but nothing captured in Sentry.  Any likely hypotheses as to what might be going wrong or what else to look for?  This is pretty hard to test, since it only occurs in production and I'm not fond of crashing my prod servers on purpose.  :smiley: Thanks!

**Comment by pkaminski** [comment]
Upon further investigation, I believe this is caused by more than one top-level exception being thrown in close succession.  To [avoid recursion](https://github.com/getsentry/raven-node/commit/6815af65d9ebda12222a94066f171480dfb29ea0) in the `patchGlobal` error handler, any exceptions received while the first exception is being captured are not processed, and the callback invoked right away.  As recommended by the docs, my callback is set to `process.exit(1)` immediately, causing the exception capture to be aborted even though the second exception didn't occur as part of the capture.

It looks like it's going to be tricky to work around this in my callback, since there's no way to distinguish between a failed capture and a skipped capture: both invoke `cb(false, err)`.  Ideally, the handler would distinguish exceptions that occurred as part of the capture, but I don't know if that's possible.

Tagging @ruimarinho since he contributed the recursion fix -- any thoughts?

**Comment by benvinegar** [comment]
@LewisJEllis – is this still relevant in 1.x?

**Comment by LewisJEllis** [comment]
Sort of - it's something I'm trying to have a more coherent answer for in what I'm doing for #257.

**Comment by LewisJEllis** [comment]
This should be addressed in #308. We don't *distinguish* multiple independent exceptions in quick succession vs further exceptions during capturing of the first one (possible but messy and brittle), but on a second exception we now wait a period of time before we let it cause us to exit, slightly longer than what we think it should ever take to capture the first exception.

Now the first exception should be captured before shutdown, even in the face of multiple uncaught exceptions in close succession, as long as the capturing does not otherwise fail or run particularly slowly. Two specific commits for reference are https://github.com/getsentry/raven-node/commit/c2aa6ff3fa9d09d7cd4a2d7e292c341e79f23aa1 and https://github.com/getsentry/raven-node/commit/47f9b107fb2da6c9400d2a8cab0588e62f6bfd21.
